### DISCUSSION TITLE
Dataset with multi target

### DISCUSSION THREAD

**Discussion by javadbayazi** [skorch-dev/skorch#954] [Q&A] [answered]
Hi, my dataset returns a tuple of the target class and some more info about the sample, for example, `y = [labels, age, gender]`. What is the best way to use the Skorch classifier?

**Comment by BenjaminBossan** [comment]
(I converted this to a Discussion since we want to test the feature, I hope you don't mind).

This question is difficult to answer with so little information. Do you have a multiclass classification problem? In that case, it should just work if you have a 2d tensor as target. Of course, the labels should be encoded, so something like this:

```python
>>> target
tensor([[0, 1, 2],
        [3, 0, 1],
        [2, 3, 0],
        ...
        [1, 2, 3],
        [0, 1, 2]])
```
 
Of course, your criterion should be able to deal with this kind of target.

If you have more details, I might be able to provide a more complete solution.

**Reply by BenjaminBossan** [reply]
@MohammadJavadD Any updates?

**Reply by javadbayazi** [reply]
Hi, @BenjaminBossan Sorry I missed this conversation. Actually, my dataloader returns the target as a tuple (i.e.` [0, 45, 1]`) you mentioned. but the criterion only needs to predict the first element. in PyTorch I would use `target[:, 0]` but here I'm getting error.

**Reply by BenjaminBossan** [reply, marked as answer]
I see, so your target contains multiple columns but you actually only need the first one. The best solution to this depends a bit on your data.

If you have control over `X` and `y`, and they are tensors or numpy arrays, this should just work: `net.fit(X, y[:, 0])`.

If you only have control over the dataset (e.g. because it comes from an external library), you could try to change its output. This could be through subclassing:

```python
class MyDataset(ExternalDataset):
    def __getitem__(self, i):
        Xi, yi = super().__getitem__()
        return Xi, yi[0]
```

Or you could wrap the other dataset:

```python
class MyDataset(torch.utils.data.Dataset):
    def __init__(self, external_dataset):
        self.external_dataset = external_dataset

    def __len__(self):
        return len(self.external_dataset)

    def __getitem__(self, i):
        Xi, yi = self.external_dataset[i]
        return Xi, yi[0]
```

(you can pass a dataset as `X` when calling `net.fit`)

If you cannot control the dataset, as a last resort, you could change the behavior of the net itself so that it ignores the extra targets when calculating the loss:

```python
class MyNet(NeuralNetClassifier):
    def get_loss(self, y_pred, y_true, **kwargs):
        y_true = y_true[:, 0]
        return super().get_loss(y_pred, y_true, **kwargs)
```

**Reply by BenjaminBossan** [reply]
@MohammadJavadD Did you manage to try any of the proposals?

**Reply by javadbayazi** [reply]
@BenjaminBossan Thank you so much! The mix of the second and third solutions works for my case!

**Reply by javadbayazi** [reply]
As I'm implementing [DANN (here)](https://arxiv.org/abs/1505.07818), I need additional information on the target. Another obstacle is how can I implement the gradient reversal layer (GRL) on top of the representation.

**Reply by BenjaminBossan** [reply]
> Another obstacle is how can I implement the gradient reversal layer (GRL) on top of the representation.

I haven't read the paper, but maybe adding a [gradient hook](https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html) can solve this. However, this is a different discussion and probably not skorch-specific, so please open a new discussion on the appropriate forum if you have more questions.
